### ISSUE TITLE
Don't ask verification code and access forbiden and apple account block

### ISSUE THREAD

**Issue by ArkaitzSKR** [fastlane/fastlane#21071]
### New Issue Checklist

- [x] Updated fastlane to the latest version
- [x] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [x] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

### Issue Description
When I try to log in to the apple account, it asks me for the username and password but not the verification code. The code arrives at the authorized apple devices but the fastlane application does not ask me for it and returns access forbiden.

##### Complete output when running fastlane, including the stack trace and command used

<details>
  <pre> 
   Please enter your Apple ID developer credentials
[11:54:57]: Apple ID Username:
apple@....
[11:56:15]: Logging in...
Password (for apple@...): **************
[11:56:45]: --------------------
[11:56:45]: fastlane init failed
[11:56:45]: --------------------
[11:56:45]: ["The request could not be completed because:", "Access forbidden"] </pre>
</details>

### Environment

<details>
  <pre> <details><summary>✅ fastlane environment ✅</summary>

### Stack

| Key                         | Value                                                      |
| --------------------------- | ---------------------------------------------------------- |
| OS                          | 13.1                                                       |
| Ruby                        | 3.1.3                                                      |
| Bundler?                    | false                                                      |
| Git                         | git version 2.34.1                                         |
| Installation Source         | /opt/homebrew/Cellar/fastlane/2.212.0/libexec/bin/fastlane |
| Host                        | macOS 13.1 (22C65)                                         |
| Ruby Lib Dir                | /opt/homebrew/Cellar/ruby@3.1/3.1.3_1/lib                  |
| OpenSSL Version             | OpenSSL 3.0.7 1 Nov 2022                                   |
| Is contained                | false                                                      |
| Is homebrew                 | true                                                       |
| Is installed via Fabric.app | false                                                      |
| Xcode Path                  | /Applications/Xcode.app/Contents/Developer/                |
| Xcode Version               | 14.2                                                       |
| Swift Version               | 5.7.2                                                      |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_US.UTF-8 | ✅ |
| LC_ALL   |             |   |
| LANGUAGE |             |   |

### fastlane files:

**No Fastfile found**

**No Appfile found**

### fastlane gems

| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.212.0 | ✅ Up-To-Date  |

### Loaded fastlane plugins:

**No plugins Loaded**

<details><summary><b>Loaded gems</b></summary>

| Gem                             | Version      |
| ------------------------------- | ------------ |
| error_highlight                 | 0.3.0        |
| did_you_mean                    | 1.6.1        |
| atomos                          | 0.1.3        |
| rexml                           | 3.2.5        |
| CFPropertyList                  | 3.0.6        |
| claide                          | 1.1.0        |
| colored2                        | 3.1.2        |
| nanaimo                         | 0.3.0        |
| xcodeproj                       | 1.22.0       |
| rouge                           | 2.0.7        |
| xcpretty                        | 0.3.0        |
| terminal-notifier               | 2.0.0        |
| unicode-display_width           | 1.8.0        |
| terminal-table                  | 1.8.0        |
| plist                           | 3.6.0        |
| public_suffix                   | 5.0.1        |
| addressable                     | 2.8.1        |
| multipart-post                  | 2.0.0        |
| word_wrap                       | 1.0.0        |
| optparse                        | 0.1.1        |
| tty-screen                      | 0.8.1        |
| tty-cursor                      | 0.7.1        |
| tty-spinner                     | 0.9.3        |
| artifactory                     | 3.0.15       |
| babosa                          | 1.0.4        |
| colored                         | 1.2          |
| highline                        | 2.0.3        |
| commander                       | 4.6.0        |
| excon                           | 0.99.0       |
| faraday-em_http                 | 1.0.0        |
| faraday-em_synchrony            | 1.0.0        |
| faraday-excon                   | 1.1.0        |
| faraday-httpclient              | 1.0.1        |
| faraday-multipart               | 1.0.4        |
| faraday-net_http                | 1.0.1        |
| faraday-net_http_persistent     | 1.2.0        |
| faraday-patron                  | 1.0.0        |
| faraday-rack                    | 1.0.0        |
| faraday-retry                   | 1.0.3        |
| ruby2_keywords                  | 0.0.5        |
| faraday                         | 1.10.3       |
| unf_ext                         | 0.0.8.2      |
| unf                             | 0.1.4        |
| domain_name                     | 0.5.20190701 |
| http-cookie                     | 1.0.5        |
| faraday-cookie_jar              | 0.0.7        |
| faraday_middleware              | 1.2.0        |
| fastimage                       | 2.2.6        |
| gh_inspector                    | 1.1.3        |
| json                            | 2.6.1        |
| mini_magick                     | 4.12.0       |
| naturally                       | 2.2.1        |
| rubyzip                         | 2.3.2        |
| security                        | 0.1.3        |
| xcpretty-travis-formatter       | 1.0.1        |
| dotenv                          | 2.8.1        |
| bundler                         | 2.3.26       |
| simctl                          | 1.6.10       |
| jwt                             | 2.7.0        |
| uber                            | 0.1.0        |
| declarative                     | 0.0.20       |
| trailblazer-option              | 0.1.2        |
| representable                   | 3.2.0        |
| retriable                       | 3.1.2        |
| mini_mime                       | 1.1.2        |
| memoist                         | 0.16.2       |
| multi_json                      | 1.15.0       |
| os                              | 1.1.4        |
| signet                          | 0.17.0       |
| googleauth                      | 1.3.0        |
| httpclient                      | 2.8.3        |
| webrick                         | 1.8.1        |
| google-apis-core                | 0.11.0       |
| google-apis-playcustomapp_v1    | 0.12.0       |
| google-apis-androidpublisher_v3 | 0.34.0       |
| google-cloud-env                | 1.6.0        |
| google-cloud-errors             | 1.3.0        |
| google-cloud-core               | 1.6.0        |
| google-apis-iamcredentials_v1   | 0.17.0       |
| google-apis-storage_v1          | 0.19.0       |
| rake                            | 13.0.6       |
| digest-crc                      | 0.6.4        |
| google-cloud-storage            | 1.44.0       |
| emoji_regex                     | 3.2.3        |
| jmespath                        | 1.6.2        |
| aws-partitions                  | 1.714.0      |
| aws-eventstream                 | 1.2.0        |
| aws-sigv4                       | 1.5.2        |
| aws-sdk-core                    | 3.170.0      |
| aws-sdk-kms                     | 1.62.0       |
| aws-sdk-s3                      | 1.119.1      |
| set                             | 1.0.2        |
| forwardable                     | 1.3.2        |
| logger                          | 1.5.0        |
| pathname                        | 0.2.0        |
| shellwords                      | 0.1.0        |
| cgi                             | 0.3.5        |
| date                            | 3.2.2        |
| timeout                         | 0.2.0        |
| stringio                        | 3.0.1        |
| securerandom                    | 0.2.0        |
| uri                             | 0.11.0       |
| openssl                         | 3.0.1        |
| digest                          | 3.1.0        |
| io-nonblock                     | 0.1.0        |
| ipaddr                          | 1.2.4        |
| io-wait                         | 0.2.1        |
| zlib                            | 2.1.1        |
| resolv                          | 0.2.1        |
| time                            | 0.2.0        |
| open-uri                        | 0.2.0        |
| mutex_m                         | 0.1.1        |
| net-http                        | 0.3.0        |
| net-protocol                    | 0.1.2        |
| ostruct                         | 0.5.2        |
| english                         | 0.7.1        |
| erb                             | 2.2.3        |
| strscan                         | 3.0.1        |
| abbrev                          | 0.1.0        |
| io-console                      | 0.5.11       |
| tempfile                        | 0.1.2        |
| delegate                        | 0.2.0        |
| fileutils                       | 1.6.0        |
| tmpdir                          | 0.1.2        |
| base64                          | 0.1.1        |
| singleton                       | 0.1.1        |
| open3                           | 0.1.1        |
| nkf                             | 0.1.1        |
| prettyprint                     | 0.1.1        |
| pp                              | 0.3.0        |
| find                            | 0.1.1        |
| yaml                            | 0.2.0        |
| psych                           | 4.0.4        |
</details>

*generated on:* **2023-02-23**
</details> </pre>
</details>

**Comment by robertsammons** [comment]
Getting same issue, be careful as it completely locked my Apple account

**Comment by maitines** [comment]
Same here since this morning, including account blocking.

**Comment by MichaVanEijk** [comment]
Same here... Having it at both version 2.209.0 as the newest version 2.212.0..

**Comment by sahilit** [comment]
Facing same issue from morning.
Yesterday it was working fine

**Comment by motrieux-thomas** [comment]
Same here since this morning :/
It doesn't seem to be related to the fastlane version.
My account was also blocked.

**Comment by mcaylus** [comment]
Same here - account blocked :-(

**Comment by vmontana** [comment]
Same symptoms here:
* Access Forbidden error seen in match and spaceauth processes.
* Happens in 2.209.1 and 2.212.0 versions.
* The account was blocked by Apple. After I unlocked it on Apple portal, the issue persists.
* I have been able to create a cookie session (spaceauth) in a machine on Dublin. The error persists in Spain (where I'm). Regional error? I don't think that it's a permanent solution since the cookie created in differents regions expires soon.

Anyone knows a possible fix?

Thanks!

**Comment by jacobsapps** [comment]
So has this just broken in the hour since my last deploy?

**Comment by WJacobsNL** [comment]
Still worked about 8 hours ago, but haven't worked for at least 4 hours since

**Comment by GabrielMassana** [comment]
Same.
No time to enter the 6-digits code and Access forbidden

**Comment by WJacobsNL** [comment]
> This seems to be due to Apple being careless somehow. You can fix it by manually changing your password and unlock your account.

I still have issues creating a new session using spaceauth then though, so my CI still can't run

**Comment by vmontana** [comment]
> This seems to be due to Apple being careless somehow. You can fix it by manually changing your password and unlock your account.
It's not a fix in my case. After unlock it and change the password, the "Access forbidden" error persists.

**Comment by theamiro** [comment]
I am in Kenya and my account was also locked. After changing the password the issue persists. It is not a geo-locked issue.

**Comment by PetterVangBraka** [comment]
I'm having the same issue with `fastlane 2.212.0`

**Comment by MichaVanEijk** [comment]
Okay, I got it working for me again now... I followed the following steps: 

1. Removing credentials from Fastlane via:
`fastlane fastlane-credentials remove --username APPLEUSERNAME`

2. Change password on Apple 

3. Adding credentials again
`fastlane fastlane-credentials add --username APPLEUSERNAME`

In the last step, it asked me to authenticate again, filling in the new password and 2FA - after that, my normal flows started working again.

**Comment by Jarutais** [comment]
The last time I have used Fastlane was on Monday and one hour ago when I tried to log in to App Store Connect, my account was also blocked and required a password change. Maybe this isn't related to Fastlane at all but some problem on Apple's side.

**Comment by ArkaitzSKR** [comment]
> 1. `fastlane fastlane-credentials remove --username APPLEUSERNAME`

This has not worked for me. I keep getting the login to the authorized phone twice with its verification code but fastlane does not ask me for it.

**Comment by atacengiz** [comment]
Have the same exact issue, getting access forbidden in terminal but my account was not blocked when I try to login dev portal. (2 hours ago everything was working)

**Comment by Devorasolomon** [comment]
I have this two tried to remove credentials and didnt help

**Comment by borderleap** [comment]
Same here...  tried that and it didn't work.

**Comment by MichaVanEijk** [comment]
Maybe I just got lucky... 😞

**Comment by gonzalo-rivas** [comment]
Sending trough Xcode fails too. It is a collapse in appstore connect itself, no "breaking api changes" for apis like fastlane.

**Comment by igorgn** [comment]
Looks like it doesn't related to account settings, as all accounts has same `Spaceship::AccessForbiddenError` issue

**Comment by xalikoutis** [comment]
Same here i changed the password but the same happens
`Could not login to App Store Connect
Please check your credentials and try again.
This could be an issue with App Store Connect,
Please try unsetting the FASTLANE_SESSION environment variable by calling 'unset FASTLANE_SESSION'
(if it is set) and re-run `fastlane spaceauth`

Exception type: Spaceship::AccessForbiddenError`

**Comment by tugsadsen** [comment]
Same here...

**Comment by itaischuftan** [comment]
same here, all of our mobile CI is down due to this

**Comment by eidson-encora** [comment]
same here :(

**Comment by kozak** [comment]
It seems like some sort of security policy change, for me this worked (on our CI): 
* unlock the account  on Apple website
* run `unset FASTLANE_SESSION` (if you are using it, might be optional) in CI 
* run `bundle exec fastlane spaceauth -u APPLE_ID` in CI to generate a new session 
* store the new FASTLANE_SESSION env var

**Comment by DovletAmanov** [comment]
> Okay, I got it working for me again now... I followed the following steps:
> 
> 1. Removing credentials from Fastlane via:
>    `fastlane fastlane-credentials remove --username APPLEUSERNAME`
> 2. Change password on Apple
> 3. Adding credentials again
>    `fastlane fastlane-credentials add --username APPLEUSERNAME`
> 
> In the last step, it asked me to authenticate again, filling in the new password and 2FA - after that, my normal flows started working again.

Same issue. Tried this solution, didn't help

**Comment by JosefBredereck** [comment]
We had the same problem and changed to using an API key that you can create with this documentation
https://docs.fastlane.tools/app-store-connect-api/

**Comment by geertbleyen** [comment]
Having the same issue here. Haven't found a work-around.

> We had the same problem and changed to using an API key that you can create with this documentation https://docs.fastlane.tools/app-store-connect-api/

Using an AppStoreConnect API key is something we are not willing to do, as there is no way to control which apps these keys can be used for and offers anonymous access to your Developer Account's AppStoreConnect.

**Comment by iamtarun** [comment]
Having same issue since morning

Could not login to App Store Connect
Please check your credentials and try again.
This could be an issue with App Store Connect,
Please try unsetting the FASTLANE_SESSION environment variable by calling 'unset FASTLANE_SESSION'
(if it is set) and re-run `fastlane spaceauth`

Exception type: Spaceship::AccessForbiddenError

No workaround (:

**Comment by geertbleyen** [comment]
Not sure what Apple is doing, but on the account affected by this issue, all app-specific-passwords also disappeared. Not sure this is caused by the account becoming locked and then unlocking it, or if this is part of the problem where Apple is messing with account passwords and authentication.

**Comment by WJacobsNL** [comment]
It seems to be a general Apple API issue with authentication. Just had another account locked for trying to access the app store through an api call...

**Comment by aebischers** [comment]
It looks like https://idmsa.apple.com added an additional security question prompt for accounts without 2FA enabled. If I understand it correctly this means that Fastlane Spaceship has to answer two security questions to successfully log in.

**Comment by geertbleyen** [comment]
@aebischers 
But the account we use that is affected by this issue, does have 2FA enabled.

**Comment by VincentDurandChewbi** [comment]
Same issues, account blocked and 2FA not working.

Exception type: Spaceship::AccessForbiddenError

Fastlane version 2.210.1

**Comment by MattKiazyk** [comment]
It's not just Fastlane having issues with Authentication. https://github.com/RobotsAndPencils/XcodesApp is having the same issue so Apple changed something around how it is checking Apple ID's and is now setting those as locked.

**Comment by gonzalo-rivas** [comment]
As i said, Xcode archiving and sending also fails. The official process is failing....

**Comment by joshdholtz** [comment]
Hey, everyone!

I've reached out to an Apple contact to see if I can get any insight into what's causing this.

Hopefully be able to give update soon 🤷‍♂️

But also... if you are able to use the App Store Connect API Key auth for any of your fastlane flows, that would probably get around this issue 🤷‍♂️

Side note: I have my two kids for the next few hours so this is not great timing 🙈 but... I'll try to do as much as I can!

**Comment by lclarkg18** [comment]
This is happening to me when running fastlane init, I'm okay with setting it up an api key but that means I can't use the TestFlight auto-set up. Anywhere I can access a sample TestFlight Fastfile? Otherwise I'll just follow the pilot instructions it looks simple

**Comment by rodrigodalri** [comment]
Same issues, account blocked and 2FA not working.

**Comment by LilaQ** [comment]
Same here, account locked, 2FA not working, forced to reset password. Right in the middle of setting up a CI, couldn't be a better timing…

**Comment by grant-mccarriagher** [comment]
We use a combination of user creds/2FA and app specific passwords, and I can confirm that our app specific passwords were blown away by this. So be sure to double check that if you're in the same boat.

**Comment by joshdholtz** [comment]
✨ Official Update 1

It looks like this is happening to everybody. I was able to reproduce it on one of my accounts 😬 

It _appears_ that Apple might have added some new "federated" login thing. I got a weird prompt but my screenshot didn't save. But I do see in my proxy logs that there are some new endpoints so I'm trying to figure out exactly what we need to change to make things start working again.

## Huge favor

🙏  If we could pause any "+1" or "this is happening to me to" that would be greatly appreciated!

I want to keep the comments for any updates from the _fastlane_ team or if anybody has done anything to fix this on their own ❤️ 

Thank you!

**Comment by timsutton** [comment]
We had an account used by CI which we normally use to fetch provisioning profiles from ADC as part of our process for performing release builds, which last night unexpectedly needed a password reset. After performing the password reset, it began working again.

I would _love_ if we could use the App Store Connect API instead of needing to depend on an ADC login flow, but we use both a regular Apple ADC account as well as an Enterprise team account, and the Enterprise team account has no support at all for the App Store Connect API. Filed FB9161884, June 2021.

The login/password flow has always felt fragile and has broken unexpectedly at various times when Apple makes backend changes, it has no SLA and mandates 2FA on new accounts. We have to treat it as a brittle part of our CI infrastructure despite it being an important part of outputting release builds for both external use and internally among hundreds of developers as well as a dedicated QA team.

**Comment by joshdholtz** [comment]
✨ Official Update 2

I got it working with SMS verification 🥳  There was a change in the API used for sending up the 2FA answers

Now testing with the other forms while my 2 year old is crawling all over me 🤷

**Comment by joshdholtz** [comment]
✨ Official Update 3

It looks like that trusted device still works without code changes but... having an existing session was causing to 💥 and lock out my account 

Once I deleted the session from `~/.fastlane/spaceship/<email>` it worked for me 🤷‍♂️ 

So... it seems like trusted device _might_ work with current _fastlane_ releases but it requires deleting that session. 

## Big Ask

Could somebody risk getting locked out again but try deleting your _fastlane_ session first? (And this will only work with trusted device and **not** SMS)

**Comment by serhiiromaniuk** [comment]
@joshdholtz Confirmed. After deletion from `~/.fastlane/spaceship/<email>` session have just generated.

**Comment by joshdholtz** [comment]
@serhiiromaniuk Thank you! Will have this next version of fastlane just destroy any existing sessions

**Comment by mailyokesh** [comment]
tried deleting the ~/.fastlane/spaceship/<email> still seeing the same issue where fastlane spaceauth -u email will hang while trying to login. I will get couple of 2FA in my trusted device and then the command fails saying forbidden access

**Comment by tmiyamon** [comment]
I deleted `~/.fastlane/spaceship/<email>` and tried again, but still got the same error `Access forbidden`.

**Comment by tk185174** [comment]
I have also deleted my email folders and still get
 "Could not login to App Store Connect"
"[!] The request could not be completed because:
	Access forbidden"
It did not generate the 2FA token or prompt for it.

**Comment by joshdholtz** [comment]
This is super weird! I have no idea why its not failing for me anymore 😱 😱 😱

**Comment by jphillips05** [comment]
I am also having the same issue, not prompting 2fa on any devices

**Comment by pierguinzani** [comment]
same here :(

**Comment by tmiyamon** [comment]
Suddenly fastlane worked well with 2fa after some trials. I have no idea why error disappeared...

**Comment by mdgeist** [comment]
I get to the point where it asks for my passcode but I never get a code sent to my phone when using `fastlane spaceauth -u xxx@xxx.com`

**Comment by yerito0** [comment]
> Suddenly fastlane worked well with 2fa after some trials. I have no idea why error disappeared...

this work for me, after some trials the sms was send for 2FA

**Comment by kigh143** [comment]
I am facing the same issue I delete my fastlane login in keychain. How do I recreate this. And I also get lots  of prompts for verification code when I run  fastlane spaceauth -u xxx@xxx.com

**Comment by joshdholtz** [comment]
✨ Official Update 4

I'm really confused at all these different behaviors now 😖 

- It works for me and a few others
- Doesn't work at all for others
- Some people aren't getting prompted codes

---

It seems that the fix in #21072 was needed for SMS but the Apple ID lockout stuff might just be something really weird going on on Apple's side? 😬

**Comment by bolo992** [comment]
It seems that there is some issue on Apple side, after a few tries, my account gets blocked, then i unblocked the account and again tried `fastlane spaceauth -u xxx@xxx.com` and it was working

**Comment by JoeSzymanski** [comment]
To add to the cases, I'm getting the 2FA code sent to my devices, but am never prompted to enter the code. It prompts a second time for the 2FA on my devices, then fails with Access Forbidden errors.

I've tried deleting the credentials and deleting the referenced folder in the .fastlane folder and still see the same behavior. My Apple account did get blocked once, but hasn't been blocked again since.

**Comment by LilaQ** [comment]
> It seems that there is some issue on Apple side, after a few tries, my account gets blocked, then i unblocked the account and again tried `fastlane spaceauth -u xxx@xxx.com` and it was working

Can confirm, worked for me too, exactly the same way.

**Comment by Devorasolomon** [comment]
I do not get an sms to recover the account

**Comment by MiguelSOliveira** [comment]
Will the fix be released in a new update?

**Comment by aebischers** [comment]
@joshdholtz We noticed that `/appleauth/auth/signin` returns **409** for Apple-IDs without 2FA enabled and asks for security questions.... Client then sends answers to questions to `/appleauth/auth/verify/questions` and finally receives code 412.
Body for questions request is:
`{"questions":[{"question":"What is the name of your favorite sports team?","answer":"XYZ-Answer1","id":147,"number":3},{"question":"What was the name of your first pet?","answer":"XYZ-Answer2","id":131,"number":1}]}`


 So https://github.com/fastlane/fastlane/blob/master/spaceship/lib/spaceship/client.rb#L516 must be adapted. Unfortunately I'm a very bad Ruby coder...

**Comment by joshdholtz** [comment]
> Will the fix be released in a new update?

Honestly... I'm not sure because this seems like issues deep within Apple's side. I've been talking to the people at https://github.com/RobotsAndPencils/XcodesApp and they were seeing the same issue here but its not reproducible and its slowly starting to work for more and more people.

I can't reproduce the locking out or anything on my side anymore.

The PR that I just merged(that hasn't been released yet) has a change for SMS 2FA but it also doesn't look like it was the cause of the issues that everybody else is seeing 😬 

I'm not giving up on the issue on my side but progress is slowed because I can't reproduce any issues anymore with any of my accounts.

**Comment by cadespres** [comment]
I deleted the spaceship folder content.
When I try `fastlane spaceauth -u xxx@xxx.com` I get `Exception type: Spaceship::AccessForbiddenError` among other things

I tried to delete the whole spaceship folder, same deal.
Hope this can help

**Comment by joshdholtz** [comment]
> @joshdholtz We noticed that `/appleauth/auth/signin` returns **409** `{ "authType" : "sa" }` for Apple-IDs without 2FA enabled and asks for security questions.... Client then sends answers to questions to `/appleauth/auth/verify/questions` and finally receives code 412. Body for questions request is: `{"questions":[{"question":"What is the name of your favorite sports team?","answer":"XYZ-Answer1","id":147,"number":3},{"question":"What was the name of your first pet?","answer":"XYZ-Answer2","id":131,"number":1}]}`
> 
> So https://github.com/fastlane/fastlane/blob/master/spaceship/lib/spaceship/client.rb#L516 must be adapted. Unfortunately I'm a very bad Ruby coder...

@aebischers Thank you for this explanation! I don't think I have an account setup like this but I will see if I can set one up and then work on fixing this 🤞

**Comment by aquacode** [comment]
The fix on the joshdholtz-fix-apple-id-2fa branch doesn't work for me. I am not getting my account blocked but I am still receiving the `Access forbidden` error.
So is the thinking that Apple has to fix this??

**Comment by MattKiazyk** [comment]
@aquacode if you go to https://developer.apple.com/download and log in, does it give you any information?

**Comment by aebischers** [comment]
@joshdholtz please have a look at your mailbox.

**Comment by aquacode** [comment]
> @aquacode if you go to https://developer.apple.com/download and log in, does it give you any information?

Logging in with the exact user via web browser I'm prompted for 2fa and all works fine. I can visit /download in developer account and it doesn't show anything special, just list of downloadable artifacts.

**Comment by joshdholtz** [comment]
> The fix on the joshdholtz-fix-apple-id-2fa branch doesn't work for me. I am not getting my account blocked but I am still receiving the `Access forbidden` error. So is the thinking that Apple has to fix this??

@aquacode I'm still unsure 😬  I can reproduce this error that you are seeing at the moment so its hard for me diagnose the root cause 😔

**Comment by aquacode** [comment]
Not sure if this is useful but since the error is still occurring for me I thought I'd share --verbose logging outputs this immediately upon supplying the password (not sure why it times out instantly):

Timeout received: 'Spaceship::AccessForbiddenError', 'Access forbidden'. Retrying after 3 seconds (remaining: 4)...
Timeout received: 'Spaceship::AccessForbiddenError', 'Access forbidden'. Retrying after 3 seconds (remaining: 3)...
Timeout received: 'Spaceship::AccessForbiddenError', 'Access forbidden'. Retrying after 3 seconds (remaining: 2)...
Timeout received: 'Spaceship::AccessForbiddenError', 'Access forbidden'. Retrying after 3 seconds (remaining: 1)...
Could not login to App Store Connect
Please check your credentials and try again.
This could be an issue with App Store Connect,
Please try unsetting the FASTLANE_SESSION environment variable by calling 'unset FASTLANE_SESSION'
(if it is set) and re-run `fastlane spaceauth`

**Comment by andrewheavin** [comment]
@aquacode - I'm still having same issue. Getting Access Forbidden

**Comment by joshdholtz** [comment]
❓ Where is everybody located that is seeing these issues? 

Another fastlane user shared a test Apple ID account with me and...
✅ worked fine for me in the United States
❌ did not work for them in Europe

It _seems_ like this could be a location based bug and we are hitting different Apple auth servers???

**Comment by andrewheavin** [comment]
I am located within the United States.

**Comment by BrunoFDeveloper** [comment]
> ❓ Where is everybody located that is seeing these issues?
> 
> Another fastlane user shared a test Apple ID account with me and... ✅ worked fine for me in the United States ❌ did not work for them in Europe
> 
> It _seems_ like this could be a location based bug and we are hitting different Apple auth servers???

I am in Brazil

**Comment by MiguelSOliveira** [comment]
Europe and does not work for me.

**Comment by fobin** [comment]
Seeing the issue in Finland, Europe

**Comment by LilaQ** [comment]
Europe, Germany, working fine (see reply above).

**Comment by demonnic** [comment]
east coast of the united states here, getting the issue on two different accounts, one uses SMS the other trusted device for 2fa.

**Comment by 99ff00** [comment]
Portugal – Access forbidden

**Comment by joshdholtz** [comment]
✨ Official Update 5

I was able to replicate the issue again by VPN-ing into Luxembourg

**Comment by cadespres** [comment]
Quebec, Canada, not working

**Comment by StringKaori** [comment]
São Paulo, Brazil, not working

**Comment by jaysonlane** [comment]
United States – not working. Have regenerated app specific password and have regenerated spaceauth session using both a device and SMS, neither of which worked.

**Comment by mlazari** [comment]
I had this issue about an hour ago. My account was blocked. I unblocked / changed the password and it worked on second try.

**Comment by tyemykuga** [comment]
São Paulo, Brazil, not working

**Comment by kigh143** [comment]
Kampala, Uganda not working

**Comment by lclarkg18** [comment]
> I had this issue about an hour ago. My account was blocked. I unblocked / changed the password and it worked on second try.

Worked for me the second time as well! (London, UK)

**Comment by Devorasolomon** [comment]
isreal

‫בתאריך יום ה׳, 23 בפבר׳ 2023 ב-21:57 מאת ‪Katende hakim. G‬‏ <‪
***@***.***‬‏>:‬

**Comment by fobin** [comment]
I first had my account locked. I switched password. And now I get different error.
```[!] Error uploading ipa file: 
 [Application Loader Error Output]: Error uploading '/var/folders/36/73kvyp892_s5n3r9ch4zwrqh0000gn/T/77bb3fae-5edb-4f20-87ae-39810c3c83af.ipa'.
[Application Loader Error Output]: Unable to upload archive. Failed to get authorization for username 'r*****'@****ify.io' and password. (
[Application Loader Error Output]: The call to the altool completed with a non-zero exit status: 1. This indicates a failure.
```

Fastlane and Organizer both fail. Would guess it's Apple's issue...

**Comment by mdgeist** [comment]
Wanted to update that my issue with my verification code not arriving had to do with my trusted device not being accessible anymore.

I had to go to https://appleid.apple.com and check what exactly was my trusted device . It was there I discovered I had it set to a VM I used in testing.  

1st attempt failed to generate a session token when running command:
`fastlane spaceauth -u xxxx@xxxx.com`

However second attempt was successful.

**Comment by grant-mccarriagher** [comment]
New York here with `Spaceship::AccessForbiddenError`.

I could be wrong, but I just want to note that this doesn't seem to have anything to do with Fastlane's 2FA flow, since the login is failing, which happens before we even know we need 2FA in the first place. Please correct me if I'm mistaken.

**Comment by joshdholtz** [comment]
✨ Official Update 6

It looks like the `Access forbidden` error returns a URL in the response body to https://support.apple.com/en-us/HT212522

Which...

```
If you see a message that says "Unable to sign you in to your Apple ID. Try again later," you've attempted to sign in to too many accounts on that device or web browser within a short period of time. You need to wait at least 24 hours before you attempt to sign in again on that device or web browser.

While there's no way to shorten this waiting period, you can continue to use other devices that are already signed in with your Apple ID. 

During this time, don't:

Sign out of any of your devices
Reset your Apple ID password
Begin an account recovery request
Doing so won't remove the message, and might further delay access to your account. 
```

It looks like Apple does not like how _often_ these Apple accounts are getting logged into 🤷‍♂️ 

## If you could help me...

If somebody who is seeing this error is able to do the following steps and see if they can log in again, the would be ❤️ 

Getting the answer to this would be super helpful for support reasons and outputting a proper error.

But it also looks like we need to do a smarter job inside of _fastlane_ of not hitting this so often 🤷‍♂️

**Comment by fobin** [comment]
Well I don't know if there is steps to try. It says **Don't** try these in Apple's documentation. At least for me it wasn't new device in any meaning as the help article states in the beginning.

**Comment by joshdholtz** [comment]
Ha! I thought it said "try these"... They really need to change that messaging

**Comment by lclarkg18** [comment]
> somebody who is seeing this error is able to do the following steps and see if they can log in again, the would be

I don't remember seeing this message but in essence this is what I had to do, I reset my password via it's account recovery system and then tried again. The second time it said "incorrect password in keychain" when trying to use the password it had remembered (because I had changed my password) but it worked anyway with my new password and asked me for the 2FA code.

Still not ideal if everyone has to reset their accounts to be able to login though

**Comment by aquacode** [comment]
Not sure anyone can explain this (as it's more an Apple question), but is one's terminal/machine considered a device? I was able to login via my browser which I would assume would be a _different_ device. In any of these "virtual" devices I do NOT see them appear under Devices on the user's appleid account. 
But now I have to wait 24 hours in order to login again from my terminal/machine??
The frustrating aspect is that I hadn't even logged in from this machine for at least a month and I needed a new FASTLANE_SESSION and now I can't retrieve one. How many logins is fastlane doing??

**Comment by joshdholtz** [comment]
❓ Are you all seeing these errors on your personal Apple IDs that are signed into your phones, Xcode, Apple TVs, etc?

I wonder if those logins are also making things weird. 

It _seems_ like because of this new rate limit stuff that it might be **highly suggested** that you use a separate account for _fastlane_ things.

Anybody have any insights or thoughts into this?

**Comment by aquacode** [comment]
that's just it, I AM using a separate account and _that's the one_ that's seemingly being rate limited, albeit I only today attempted to login with it after a month hiatus.

**Comment by joshdholtz** [comment]
That is _very_ suspicious that these are _fastlane_ specific accounts 👀 

I honestly wouldn't have guessed that

**Comment by 99ff00** [comment]
I use account for `fastlane` only. It suddenly started today.

**Comment by mdgeist** [comment]
I tested with my own account but use a different dedicated account for CI with Fastlane.

**Comment by aquacode** [comment]
> ❓ Are you all seeing these errors on your personal Apple IDs that are signed into your phones, Xcode, Apple TVs, etc?
> 
> I wonder if those logins are also making things weird.
> 
> It _seems_ like because of this new rate limit stuff that it might be **highly suggested** that you use a separate account for _fastlane_ things.
> 
> Anybody have any insights or thoughts into this?

I am ONLY seeing this on my CI-specific account, only started today, and first occurred on a CircleCI workflow. Then, I started having the issue on my terminal for the same account. All my personal Apple IDs seem fine i.e. not getting this weird "too many logins" problem (hoping I didn't jinx myself).

**Comment by grant-mccarriagher** [comment]
Also using a dedicated CI service account here. In our case, we only do a spaceship login twice per day to grab a session token, which is then used by our other CI jobs (which are on different machines in the same network location) instead of logging in each time.

**Comment by mieliespoor** [comment]
> Also using a dedicated CI service account here. In our case, we only do a spaceship login twice per day to grab a session token, which is then used by our other CI jobs (which are on different machines in the same network location) instead of logging in each time.

Without wanting to derail this thread, I would like to find out more here as this would help us as well.

On the main issue, we saw this late this afternoon, but will test again to provide feedback. We also use a dedicated account for our CI environment that nobody should be logging into via browser etc.

**Comment by joshdholtz** [comment]
✨ Official Update 7

So based on this feedback, some new headers I discovered earlier, and talking with people over at Appfigures... it looks like Apple implemented [Hashcash](http://www.hashcash.org/) in the API and that is what is causing this

I'm going to be looking into this further and _should_ have a fix out for this... hopefully 🤞 

I don't have an ETA but I will dedicate this evening and I should be able to block off a lot of time tomorrow for this.

This Hashcash looks like to help them determine denial of service stuff / too many signins from new devices

So...

Stay tuned and thank you all for your patience 🚀

**Comment by dhinakg** [comment]
Looks like you already figured out that Hashcash was added, but figured I'd post my notes anyway

---

I can't answer why this seems conditionally broken for people, but I think I've found a way to fix this.

Doing some testing by logging into https://developer.apple.com/download/all/?q=Safari through a browser and messing with the headers, it looks like Apple introduced the `X-APPLE-HC` header. Googling shows this seems to be a form of [Hashcash](http://www.hashcash.org/), except without the `rand` part and with the date format changed.

<details>
<summary>More (Apple) hashcash details</summary>

The hashcash is in the format `ver:bits:date:res:ext::counter` (ie. `1:12:20230223210829:6645198e569a298d0cb630f6341e6ef3::679`), where:

* `ver`: `1`
* `bits`: minimum number of leading zero bits in the hashcash (see below on how it is determined)
* `date`: current UTC time: `YYYYMMDDHHMMSS`
* `res`: resource string (see below on how it is determined)
* `ext`: extensions (unused by Apple)
* `counter`: just a counter of how many iterations were required (doesn't really matter)

At a high level, to "mint" a hashcash, construct the string `ver:bits:date:res:ext::counter`, hash it (SHA1), and check if it has enough zero bits at the start (our sample hashcash starts with 13 zero bits). Otherwise, increment counter and try again. Once you get a hash that has enough zero bits, you have your hashcash.

</details>

Now, in order to get the minimum bit count (it seems to vary) and the resource string, you have two options:
* Make a normal GET request to `https://idmsa.apple.com/appleauth/auth/signin` (the website uses the values from this request)
* Make a POST request to federate (I have no clue what this is for, but the website does it when you enter your email): `https://idmsa.apple.com/appleauth/auth/federate`. Payload is JSON, `{"accountName": "EMAIL", "rememberMe": true/false}`

In both cases, you'll get `X-Apple-HC-Challenge` and `X-Apple-HC-Bits` in the response headers. Use these to generate the hashcash, pass it as `X-APPLE-HC` and you should be set.

I got my (personal) Apple ID locked by my own code (had it running in a loop, after getting this error 3 times in a very short period it got locked) and had to change my password to unlock it. I still get `-36607` from https://github.com/RobotsAndPencils/XcodesApp and from my own code without adding this, but as soon as I add this it works again.

Hope this helps

**Comment by iliakukharev** [comment]
Have you experienced account locks only when attempting to sign in with both your password and 2FA code? Once the session has been established, is everything functioning normally?

**Comment by miskelma** [comment]
Have you experienced account locks only when attempting to sign in with both your password and 2FA code? Once the session has been established, is everything functioning normally?

**Comment by jaysonlane** [comment]
Mine has magically started working with no changes.

**Comment by BrainstemStudio** [comment]
Seems to be outages that are being attended to https://developer.apple.com/system-status/. Might be a wait-and-see situation before making drastic changes.

**Comment by pelinche** [comment]
> Seems to be outages that are being attended to https://developer.apple.com/system-status/. Might be a wait-and-see situation before making drastic changes.

apple's status says it has no more problems, but fastlane is still unable to get the certificates

![image](https://user-images.githubusercontent.com/71408972/221063241-42ec654b-01d4-44a7-b110-33ec5af302a6.png)

**Comment by pH-7** [comment]
> ✨ Official Update 7
> 
> So based on this feedback, some new headers I discovered earlier, and talking with people over at Appfigures... it looks like Apple implemented [Hashcash](http://www.hashcash.org/) in the API and that is what is causing this
> 
> I'm going to be looking into this further and _should_ have a fix out for this... hopefully 🤞
> 
> I don't have an ETA but I will dedicate this evening and I should be able to block off a lot of time tomorrow for this.
> 
> This Hashcash looks like to help them determine denial of service stuff / too many signins from new devices
> 
> So...
> 
> Stay tuned and thank you all for your patience 🚀

We highly appreciate this huge effort. That means we would need to wait for an update from Fastlane to have it working...?

**Comment by fossage** [comment]
While this gets fixed, we have been getting by setting `MATCH_READONLY=true` when using match. This will force it to only  grab the profiles/certs that already exist in the associated git repo rather than trying to check if it needs to create new ones by reaching out to Apple.

**Comment by joshdholtz** [comment]
@pH-7 Yup yup, you have to wait if you using Apple ID auth but if you switch to use the App Store Connect API, you can use things right away 😊 

Which... ASC API key is so much more stable so that is always my recommendation if you can use it

**Comment by joshdholtz** [comment]
✨ Official Update 8

I have a PR #21073 ready for review and some testers 😇 

**Summary:** It adds a new `X-APPLE-HC` header to the auth flow that is simple to generate but not documented. 

We got some awesome help by our friends at [Appfigures](https://appfigures.com) who were also working on solving this issue today ❤️ 

### Testing Steps

Update `Gemfile` and run `bundle install`, `bundle update fastlane`, or `bundle update`

```rb
gem "fastlane", :git => "https://github.com/fastlane/fastlane.git", :branch => "joshdholtz-implement-hashcash"
```

🙏  Please leave any question or issues about the PR on the PR and _not_ on this thread

But... you can 👍 this comment and comment if PR worked for you in this thread

Just trying to make sure that this thread doesn't become a huge conversation about the PR 😇

**Comment by ChristineWasike** [comment]
> @pH-7 Yup yup, you have to wait if you using Apple ID auth but if you switch to use the App Store Connect API, you can use things right away 😊
> 
> Which... ASC API key is so much more stable so that is always my recommendation if you can use it

I'm running FastLane on flutter. How would I switch from Apple ID auth to the App Store Connect API? Thanks! 🤗
My current FastLane actions in my lane (beta) are just three:
```
desc "Push a new beta build to TestFlight"
  lane :beta do
    increment_build_number(xcodeproj: "Runner.xcodeproj")
    build_app(workspace: "Runner.xcworkspace", scheme: "Runner")
    upload_to_testflight(app_identifier: "xxx",
    itc_provider:"xxx" )
    # slack(message: "Successfully distributed a new beta build")
  end
```

**Comment by joshdholtz** [comment]
✨ Official Update 8

I have a PR #21073 ready for review and some testers 😇 

**Summary:** It adds a new `X-APPLE-HC` header to the auth flow that is simple to generate but not documented. 

We got some awesome help by our friends at [Appfigures](https://appfigures.com) who were also working on solving this issue today ❤️ 

## Testing Steps

Update `Gemfile` and run `bundle install`, `bundle update fastlane`, or `bundle update`

```rb
gem "fastlane", :git => "https://github.com/fastlane/fastlane.git", :branch => "joshdholtz-implement-hashcash"
```

🙏  Please leave any question or issues about the PR on the PR and _not_ on this thread

But... you can 👍 this comment and comment if PR worked for you in this thread

Just trying to make sure that this thread doesn't become a huge conversation about the PR 😇

**Comment by yanniks** [comment]
PR works great for us (location Germany). Thanks!

**Comment by shivkumarsharma** [comment]
Same symptoms here:

Access Forbidden error seen in match processes and CI workflow
Happens in 2.209.1, 2.210.1 and 2.212.0 versions.
Yesterday it works fine.
The account was blocked by Apple. After I unlocked it on Apple portal, the issue persists.

Anyone knows a possible fix?

Thanks!

**Comment by mmaheo** [comment]
This works for me!! Huge thanks @joshdholtz ! 👌

**Comment by ChristineWasike** [comment]
> ✨ Official Update 8
> 
> I have a PR #21073 ready for review and some testers 😇
> 
> **Summary:** It adds a new `X-APPLE-HC` header to the auth flow that is simple to generate but not documented.
> 
> We got some awesome help by our friends at [Appfigures](https://appfigures.com) who were also working on solving this issue today ❤️
> 
> ## Testing Steps
> Update `Gemfile` and run `bundle install`, `bundle update fastlane`, or `bundle update`
> 
> ```ruby
> gem "fastlane", :git => "https://github.com/fastlane/fastlane.git", :branch => "joshdholtz-implement-hashcash"
> ```
> 
> 🙏 Please leave any question or issues about the PR on the PR and _not_ on this thread
> 
> But... you can 👍 this comment and comment if PR worked for you in this thread
> 
> Just trying to make sure that this thread doesn't become a huge conversation about the PR 😇
> 
> I need to sleep now so hoping to wake up to a bunch of "This works for me!" in the morning so I can push out a new release with this fix 😊

I'm not quite sure how to go about implementing this on my end. Might you break this down further if possible? Or perhaps someone who's also tried this and it works. Thanks.

**Comment by WJacobsNL** [comment]
PR also fixes the issues on my end! Thanks @joshdholtz

**Comment by mmaheo** [comment]
@joshdholtz Your fix seems to fix the issue, but now I got a new error 

>[!] Error uploading ipa file:
 [Application Loader Error Output]: Error uploading '/var/folders/8_/2vd36wqn2sq5y9skr7gwjcyw0000gn/T/458535fe-4479-498b-b245-138bb5384ec6.ipa'.
[Application Loader Error Output]: Unable to upload archive. Failed to get authorization for username '<MY USERNAME>' and password. (
[Application Loader Error Output]: The call to the altool completed with a non-zero exit status: 1. This indicates a failure.

**Comment by limzb** [comment]
PR works us, thank you, big time!

**Comment by ChristineWasike** [comment]
> @joshdholtz Your fix seems to fix the login issue to App Store Connect, but now I got a new error when upload a build
> 
> > [!] Error uploading ipa file:
> > [Application Loader Error Output]: Error uploading '/var/folders/8_/2vd36wqn2sq5y9skr7gwjcyw0000gn/T/458535fe-4479-498b-b245-138bb5384ec6.ipa'.
> > [Application Loader Error Output]: Unable to upload archive. Failed to get authorization for username '' and password. (
> > [Application Loader Error Output]: The call to the altool completed with a non-zero exit status: 1. This indicates a failure.

How did you go about setting up the PR? Are you building for iOS? I need some guidance getting the PR set up on my end to begin with.

**Comment by limzb** [comment]
> > ✨ Official Update 8
> > I have a PR #21073 ready for review and some testers 😇
> > **Summary:** It adds a new `X-APPLE-HC` header to the auth flow that is simple to generate but not documented.
> > We got some awesome help by our friends at [Appfigures](https://appfigures.com) who were also working on solving this issue today ❤️
> > ## Testing Steps
> > Update `Gemfile` and run `bundle install`, `bundle update fastlane`, or `bundle update`
> > ```ruby
> > gem "fastlane", :git => "https://github.com/fastlane/fastlane.git", :branch => "joshdholtz-implement-hashcash"
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 🙏 Please leave any question or issues about the PR on the PR and _not_ on this thread
> > But... you can 👍 this comment and comment if PR worked for you in this thread
> > Just trying to make sure that this thread doesn't become a huge conversation about the PR 😇
> > I need to sleep now so hoping to wake up to a bunch of "This works for me!" in the morning so I can push out a new release with this fix 😊
> 
> I'm not quite sure how to go about implementing this on my end. Might you break this down further if possible? Or perhaps someone who's also tried this and it works. Thanks.

To use the fix from new PR branch you just have to replace the the `gem "fastlane"` line in your Gemfile as such:
`
# gem "fastlane"
gem "fastlane", :git => "https://github.com/fastlane/fastlane.git", :branch => "joshdholtz-implement-hashcash"
`

**Comment by ChristineWasike** [comment]
> > > ✨ Official Update 8
> > > I have a PR #21073 ready for review and some testers 😇
> > > **Summary:** It adds a new `X-APPLE-HC` header to the auth flow that is simple to generate but not documented.
> > > We got some awesome help by our friends at [Appfigures](https://appfigures.com) who were also working on solving this issue today ❤️
> > > ## Testing Steps
> > > Update `Gemfile` and run `bundle install`, `bundle update fastlane`, or `bundle update`
> > > ```ruby
> > > gem "fastlane", :git => "https://github.com/fastlane/fastlane.git", :branch => "joshdholtz-implement-hashcash"
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > 🙏 Please leave any question or issues about the PR on the PR and _not_ on this thread
> > > But... you can 👍 this comment and comment if PR worked for you in this thread
> > > Just trying to make sure that this thread doesn't become a huge conversation about the PR 😇
> > > I need to sleep now so hoping to wake up to a bunch of "This works for me!" in the morning so I can push out a new release with this fix 😊
> > 
> > 
> > I'm not quite sure how to go about implementing this on my end. Might you break this down further if possible? Or perhaps someone who's also tried this and it works. Thanks.
> 
> To use the fix from new PR branch you just have to replace the the `gem "fastlane"` line in your Gemfile as such: `
> 
> # gem "fastlane"
> gem "fastlane", :git => "https://github.com/fastlane/fastlane.git", :branch => "joshdholtz-implement-hashcash" `

Thanks! That was clear ✨

**Comment by geertbleyen** [comment]
For me it started working again on 2.212.0. No idea...

**Comment by WJacobsNL** [comment]
> @joshdholtz Your fix seems to fix the login issue to App Store Connect, but now I got a new error when upload a build
> 
> > [!] Error uploading ipa file:
> > [Application Loader Error Output]: Error uploading '/var/folders/8_/2vd36wqn2sq5y9skr7gwjcyw0000gn/T/458535fe-4479-498b-b245-138bb5384ec6.ipa'.
> > [Application Loader Error Output]: Unable to upload archive. Failed to get authorization for username '' and password. (
> > [Application Loader Error Output]: The call to the altool completed with a non-zero exit status: 1. This indicates a failure.

You probably need to generate a new app specific password, these get whiped when your account gets locked (at least that was the case on my end)

**Comment by ChristineWasike** [comment]
Still getting the same access forbidden error.
<img width="564" alt="Screenshot 2023-02-24 at 11 12 50" src="https://user-images.githubusercontent.com/28866174/221127196-3c374ae1-150a-4505-9863-61fa62df802e.png">

Updated GemFile paths:
`ios/Gemfile`
`Gemfile`
Update my `Gemfile`  to this:

```
source "https://rubygems.org"

# gem "fastlane"
gem "fastlane", :git => "https://github.com/fastlane/fastlane.git", :branch => "joshdholtz-implement-hashcash"

```

**Comment by geertbleyen** [comment]
@WJacobsNL 
In case you are/were using an app-specific-password somewhere for Fastlane, due the issue yesterday my CI account's app-specific-password's were all wiped and I had to create a new one.
This may explain your issue (I had it too because I forgot to update the app-specific-password) in our service connection for Azure Devops that uses fastlane.

**Comment by geertbleyen** [comment]
@ChristineWasike Have you recreated your spaceauth session after the account got locked and unlocked. Normally the password is different now and the spaceauth session you had before won't be valid anymore.

**Comment by limzb** [comment]
@ChristineWasike also just in case this step were missed, did you run bundle install after changing your Gemfile?

**Comment by RalfK92** [comment]
PR works great for us. Greetings from Austria. Thanks!

**Comment by TakealotThabang** [comment]
Still getting the auth error when running fastlane spaceauth -u after updating the Gemfile

**Comment by mmaheo** [comment]
> Still getting the same access forbidden error. <img alt="Screenshot 2023-02-24 at 11 12 50" width="564" src="https://user-images.githubusercontent.com/28866174/221127196-3c374ae1-150a-4505-9863-61fa62df802e.png">
> 
> Updated GemFile paths: `ios/Gemfile` `Gemfile` Update my `Gemfile` to this:
> 
> ```
> source "https://rubygems.org"
> 
> # gem "fastlane"
> gem "fastlane", :git => "https://github.com/fastlane/fastlane.git", :branch => "joshdholtz-implement-hashcash"
> ```

@ChristineWasike  Have you run `bundle fastlane update`?

**Comment by mmaheo** [comment]
> > @joshdholtz Your fix seems to fix the login issue to App Store Connect, but now I got a new error when upload a build
> > > [!] Error uploading ipa file:
> > > [Application Loader Error Output]: Error uploading '/var/folders/8_/2vd36wqn2sq5y9skr7gwjcyw0000gn/T/458535fe-4479-498b-b245-138bb5384ec6.ipa'.
> > > [Application Loader Error Output]: Unable to upload archive. Failed to get authorization for username '' and password. (
> > > [Application Loader Error Output]: The call to the altool completed with a non-zero exit status: 1. This indicates a failure.
> 
> You probably need to generate a new app specific password, these get whiped when your account gets locked (at least that was the case on my end)

Yep you right, it's working now, thanks!
So the PR works for me, thanks 😉

**Comment by pelinche** [comment]
Here it is working using fastlane 2.212.0.

I only removed stored credentials on fastlane:

`fastlane fastlane-credentials remove --username <your_apple_username>`

and, adding again with the command

`fastlane fastlane-credentials add --username <your_apple_username>`

Informing the apple developer password. The One Time Token was sended to my mobile number, and the fastlane was waiting on prompt to insert the OTP. I inserted the code received, and all is working.

**Comment by fobin** [comment]
I still get error like this:

```[10:52:56]: Error uploading ipa file: 
 [Application Loader Error Output]: Error uploading '/var/folders/36/73kvyp892_s5n3r9ch4zwrqh0000gn/T/e2e28548-be18-4a39-ae77-a0b97276ec35.ipa'.
[Application Loader Error Output]: Unable to upload archive. Failed to get authorization for username 'r****@r****.io' and password. (
[Application Loader Error Output]: The call to the altool completed with a non-zero exit status: 1. This indicates a failure.```

fastlane 2.212.0 is the version. I managed to upload with Organizer. Notice I'm not using app specific password. I deleted my credentials and it correctly asked for pw and OTP but after that upload fails.

**Comment by vmontana** [comment]
"bundle exec fastlane spaceauth -u <APPLE_ID>" is working prefectly now using the new branch from Spain! good job!

**Comment by kigh143** [comment]
> Here it is working using fastlane 2.212.0 (default version, without use other branch)
> 
> I only removed stored credentials on fastlane:
> 
> `fastlane fastlane-credentials remove --username <your_apple_username>`
> 
> and, adding again with the command
> 
> `fastlane fastlane-credentials add --username <your_apple_username>`
> 
> Informing the apple developer password. The One Time Token was sended to my mobile number, and the fastlane was waiting on prompt to insert the OTP. I inserted the code received, and all is working.

Hey @pelinche , I get a prompt on my phone for code, how do I change to just receive otp on my phone.

**Comment by maitines** [comment]
Works perfectly with the joshdholtz branch, tried several methods and everything works fine. Great job.

![Captura de pantalla 2023-02-24 a las 10 34 05](https://user-images.githubusercontent.com/44603169/221154849-0b12d109-77fa-4882-93eb-b58baca2dcca.png)

**Comment by AmauriC** [comment]
Good morning Josh!
Working from France, thanks ❤️

**Comment by ChristineWasike** [comment]
> Works perfectly with the joshdholtz branch, tried several methods and everything works fine. Great job.
> 
> ![Captura de pantalla 2023-02-24 a las 10 34 05](https://user-images.githubusercontent.com/44603169/221155905-07bc2627-f337-4e29-a054-99cdf28c796c.png)

Nice, so I got the nod on this, but I now get a new error here:
```
[altool]     "Error Domain=ITunesConnectionOperationErrorDomain Code=-19000 \"Sign in with the app-specific password you generated. If you forgot the app-specific password or need to create a new one, go to appleid.apple.com\" UserInfo={NSLocalizedRecoverySuggestion=Sign in with the app-specific password you generated. If you forgot the app-specific password or need to create a new one, go to appleid.apple.com, NSLocalizedDescription=Sign in with the app-specific password you generated. If you forgot the app-specific password or need to create a new one, go to appleid.apple.com, NSLocalizedFailureReason=App Store operation failed.}"
```

**Comment by ChristineWasike** [comment]
> I still get error like this:
> 
> ```
>  [Application Loader Error Output]: Error uploading '/var/folders/36/73kvyp892_s5n3r9ch4zwrqh0000gn/T/e2e28548-be18-4a39-ae77-a0b97276ec35.ipa'.
> [Application Loader Error Output]: Unable to upload archive. Failed to get authorization for username 'r****@r****.io' and password. (
> [Application Loader Error Output]: The call to the altool completed with a non-zero exit status: 1. This indicates a failure.
> ```
> 
> fastlane 2.212.0 is the version. I managed to upload with Organizer. Notice I'm not using app specific password. I deleted my credentials and it correctly asked for pw and OTP but after that upload fails.

Same issue for me

**Comment by AmauriC** [comment]
@ChristineWasike Previous app specific passwords are now invalid, generate a new one from https://appleid.apple.com/

**Comment by maitines** [comment]
@ChristineWasike In my case, when apple blocked the account it also deleted app-specific password and I have had to create it again, is it possible that this is the case?

**Comment by pelinche** [comment]
> 

do you have 2 factor authentication turned on, with your phone number linked to apple account?

**Comment by pelinche** [comment]
> Hey @pelinche , I get a prompt on my phone for code, how do I change to just receive otp on my phone.

@kigh143 do you have 2 factor authentication turned on, with your phone number linked to apple account?

**Comment by ArkaitzSKR** [comment]
> ✨ Official Update 8
> 
> I have a PR #21073 ready for review and some testers 😇
> 
> **Summary:** It adds a new `X-APPLE-HC` header to the auth flow that is simple to generate but not documented.
> 
> We got some awesome help by our friends at [Appfigures](https://appfigures.com) who were also working on solving this issue today ❤️
> ## Testing Steps
> 
> Update `Gemfile` and run `bundle install`, `bundle update fastlane`, or `bundle update`
> 
> ```ruby
> gem "fastlane", :git => "https://github.com/fastlane/fastlane.git", :branch => "joshdholtz-implement-hashcash"
> ```
> 
> 🙏 Please leave any question or issues about the PR on the PR and _not_ on this thread
> 
> But... you can 👍 this comment and comment if PR worked for you in this thread
> 
> Just trying to make sure that this thread doesn't become a huge conversation about the PR 😇
> 
> I need to sleep now so hoping to wake up to a bunch of "This works for me!" in the morning so I can push out a new release with this fix 😊

This has not worked for me. We use the apple device code and in this case it continues to give me access forbidden.

😭 - [ ] Fix for Apple device code

**Comment by maitines** [comment]
> 

First I deleted the cookie in .fastlane/spaceship and then I got the new one by running "bundle exec fastlane spaceauth -u <email> Then I tried uploading to testflight and apple store and everything is fine.

**Comment by MiguelSOliveira** [comment]
Everything is working for me now, since using the branch.

**Comment by fobin** [comment]
Got it now working changing auth from pw to api key.

**Comment by BrainstemStudio** [comment]
Working with PR https://github.com/fastlane/fastlane/pull/21073

**Comment by sjors98** [comment]
PR https://github.com/fastlane/fastlane/pull/21073 works fine for me! Had to delete the cookie and set a new App Specific Password but now it is working again.
Thanks for the quick solution and hard work @joshdholtz!

**Comment by aquacode** [comment]
Using this: `gem "fastlane", :git => "https://github.com/fastlane/fastlane.git", :branch => "joshdholtz-implement-hashcash"` worked instantly.
Thanks @joshdholtz !

**Comment by kkaun** [comment]
Switched to the `branch => "joshdholtz-implement-hashcash"` and re-created the app-specific password - and it finally worked for us. Thanks @joshdholtz !

**Comment by stack** [comment]
I can confirm the branch worked for me when using `bundle exec fastlane codesigning`

**Comment by Jarutais** [comment]
My archive uploads worked doing the following

1. Switched the gemfile to `branch => "joshdholtz-implement-hashcash"`
2. Re-created app-specific password and updated it in the environment
3. Deleted the file at `.fastlane/spaceship`
4. Ran my lane as usual

**Comment by joaobzao** [comment]
> My archive uploads worked doing the following
> 
> 1. Switched the gemfile to `branch => "joshdholtz-implement-hashcash"`
> 2. Re-created app-specific password and updated it in the environment
> 3. Deleted the file at `.fastlane/spaceship`
> 4. Ran my lane as usual

works, great job @joshdholtz

**Comment by leoniralves** [comment]
Great job @joshdholtz!!! ❤️ 
Your PR worked for me. 👏

**Comment by PrimeLantern** [comment]
I can confirm that this branch worked for me 👍 
`gem "fastlane", :git => "https://github.com/fastlane/fastlane.git", :branch => "joshdholtz-implement-hashcash"`
I also had to create a new app-specific password and set it in my CI environment.

Phenomenal work @joshdholtz and thank You so much!
Greetings from 🇨🇿

**Comment by joshdholtz** [comment]
✨ Official Update 9 (I lost count)

Will be merging the PR and preparing a new version! New version _should_ be out within an hour or so if nothing breaks during the release process 🙈 

Thank you to everyone who participated in the conversation to help us diagnose the problem and thank you all for your patience ❤️

**Comment by hervouinc** [comment]
I also confirm that the PR worked for me (2FA with phone number code).

Thanks A LOT  @joshdholtz ! 🚀 

What's even better with this whole situation is that it was a great opportunity to move from 2FA / Session to AppStore Connect API Key.
At the time we first implemented Fastlane -> AppStore Connect authentication, some key actions/tools we were using were not yet available using AppStore Connect API Key.
Now it's all available.

The integration is very smooth compared to 2FA. 
I couldn't recommend more to anyone still having trouble here to user AppStore Connect API Key instead of 2FA.

Cheers 🍻

**Comment by squallstar** [comment]
@joshdholtz amazing work! When is version 2.212.1 going to be available via [rubygems](https://rubygems.org/gems/fastlane)?

**Comment by joshdholtz** [comment]
> @joshdholtz amazing work! When is version 2.212.1 going to be available via [rubygems](https://rubygems.org/gems/fastlane)?

Once I figure out why my local release process is failing 🙈 But sooooooooooon

**Comment by joshdholtz** [comment]
✨ Official Update 10

Okay! So `2.212.1` is out on RubyGems and is waiting for approval on Homebrew (probably out sometime today)

Release 👉  https://github.com/fastlane/fastlane/releases/tag/2.212.1

Hope this all works for everyone now! And... its been a pleasure talking to you all but let's not meet like this again 😉

**Comment by anilgr** [comment]
Thank you ! @joshdholtz

**Comment by StringKaori** [comment]
i can't update it, when i run `fastlane update_fastlane` it says `Updating fastlane from 2.212.0 to 2.212.1... 🚀`, but when i run `fastlane --version` it says i'm still on 2.212.0

**Comment by D3icidal** [comment]
> i can't update it, when i run `fastlane update_fastlane` it says `Updating fastlane from 2.212.0 to 2.212.1... 🚀`, but when i run `fastlane --version` it says i'm still on 2.212.0

@StringKaori  Is that a gem or homebrew installation? `fastlane env` can tell you. 
Homebrew isn't ready yet but should be soon. 

>Okay! So 2.212.1 is out on RubyGems and is waiting for approval on Homebrew (probably out sometime today)

**Comment by justincbeck** [comment]
FWIW, `fastlane update_fastlane` worked for me.

**Comment by StringKaori** [comment]
> > i can't update it, when i run `fastlane update_fastlane` it says `Updating fastlane from 2.212.0 to 2.212.1... 🚀`, but when i run `fastlane --version` it says i'm still on 2.212.0
> 
> @StringKaori Is that a gem or homebrew installation? `fastlane env` can tell you. Homebrew isn't ready yet but should be soon.
> 
> > Okay! So 2.212.1 is out on RubyGems and is waiting for approval on Homebrew (probably out sometime today)
 
@D3icidal homebrew

**Comment by D3icidal** [comment]
@StringKaori Actually can you try again? I just tried it again 10 minutes ago and brew was able to install it finally

**Comment by StringKaori** [comment]
it worked now, thanks

**Comment by deggertsen** [comment]
It's available on homebrew now. Thank you!

**Comment by courtsimas** [comment]
For some reason fastlane 2.212.1 didn't fix it for me...

**Comment by carlosriveroib** [comment]
`fastlane update_fastlane` updated the version but still not working

<img width="865" alt="Captura de pantalla 2023-02-24 a las 21 33 42" src="https://user-images.githubusercontent.com/9528363/221286188-7e0b0864-11fa-4bfb-8898-93c14cb918d7.png">

**Comment by justincbeck** [comment]
Confirmed. Thanks, so much!

**Comment by conradpronto** [comment]
> I also confirm that the PR worked for me (2FA with phone number code).
> 
> Thanks A LOT @joshdholtz ! 🚀
> 
> What's even better with this whole situation is that it was a great opportunity to move from 2FA / Session to AppStore Connect API Key. At the time we first implemented Fastlane -> AppStore Connect authentication, some key actions/tools we were using were not yet available using AppStore Connect API Key. Now it's all available.
> 
> The integration is very smooth compared to 2FA. I couldn't recommend more to anyone still having trouble here to use AppStore Connect API Key instead of 2FA.
> 
> Cheers 🍻

Do you know where we can see the scope of actions that are supported by the AppStore Connect API key?  for example, can we use it create Apple Pass Type Identifiers and create NFC Certificates.

**Comment by mailyokesh** [comment]
changing to use api_key did the trick. Also api_key seems to be supporting most of the features now and we dont have to rotate the session every 30 days.

**Comment by mdgeist** [comment]
> changing to use api_key did the trick. Also api_key seems to be supporting most of the features now and we dont have to rotate the session every 30 days.

Just curious where did you use `api_key`?

**Comment by mdgeist** [comment]
> NFC Certificates

Check out Apple's API here:
https://developer.apple.com/documentation/appstoreconnectapi/certificatetype

These are the certificates you can generate with ASC API.

**Comment by superandrew213** [comment]
@courtsimas how did you unlock your Apple ID?

**Comment by courtsimas** [comment]
@superandrew213 I logged into appleid and it notified me there. Had to jump through a few screens to unlock it.

**Comment by ArkaitzSKR** [comment]
> ✨ Official Update 10
> 
> Okay! So `2.212.1` is out on RubyGems and is waiting for approval on Homebrew (probably out sometime today)
> 
> Release 👉 https://github.com/fastlane/fastlane/releases/tag/2.212.1
> 
> Hope this all works for everyone now! And... its been a pleasure talking to you all but let's not meet like this again 😉

Thank you @joshdholtz !

**Comment by hervouinc** [comment]
https://docs.fastlane.tools/app-store-connect-api/

**Comment by hervouinc** [comment]
@conradpronto 
> Do you know where we can see the scope of actions that are supported by the AppStore Connect API key? for example, can we use it create Apple Pass Type Identifiers and create NFC Certificates.
Here is the scope of supported Fastlane actions/tools with API Key:
https://docs.fastlane.tools/app-store-connect-api/

**Comment by Lakston** [comment]
updating to latest did not solve it for me

**Comment by carlosriveroib** [comment]
> updating to 2.212.1 did not solve it for me

That happened to me. You have to go to appleId and create a new application password

**Comment by Lakston** [comment]
what do you mean by `application password` ?

**Comment by mlazari** [comment]
@Lakston I think he meant "application specific password" - https://support.apple.com/en-us/HT204397

**Comment by Lakston** [comment]
Ok our developpers account was blocked as mentionned by other people in this issue, I had to unblock it and change the password

**Comment by mailyokesh** [comment]
> > changing to use api_key did the trick. Also api_key seems to be supporting most of the features now and we dont have to rotate the session every 30 days.
> 
> Just curious where did you use `api_key`?

instruction on how to generate apikey and then put it as a json file 
Here is my fastfile 

default_platform(:ios)

platform :ios do
  desc "Push a new beta build to TestFlight"
  lane :beta do
    upload_to_testflight(api_key_path: "fastlane/somefile.json",distribute_external: true,  changelog:"02-27-2023 09:45:10")
  end
end

contents of somefile.json placed inside fastlane folder 
{
  "key_id": "your key id",
  "issuer_id": "your issuer id",
  "key": "-----BEGIN PRIVATE KEY-----\n\n-----END PRIVATE KEY-----",
  "duration": 1200,
  "in_house": false
}
